### PR TITLE
Pass penalties and seed params in v0 model runner 

### PR DESCRIPTION
**Pull request by sraizada-tt**
(https://github.com/tenstorrent/tt-metal/issues/32212)

Metal changes: https://github.com/tenstorrent/tt-metal/pull/32807

vllm nightly: https://github.com/tenstorrent/tt-metal/actions/runs/19770818868

### CODE REVIEW

File: vllm/worker/tt_model_runner.py [hunk 1]
```diff
@@ -568,6 +612,17 @@ def prepare_model_input(
                                     dtype=torch.int32,
                                     device="cpu")
                     ])
+                if prompt_tokens_tensor is not None:
+                    pad_rows = input_tokens.shape[
+                        0] - prompt_tokens_tensor.shape[0]
+                    if pad_rows > 0:
+                        prompt_tokens_tensor = torch.cat([
+                            prompt_tokens_tensor,
+                            torch.zeros(pad_rows,
```
Comment (skhorasganiTT) [inline vllm/worker/tt_model_runner.py]: Can use batch_pad_len instead of recalculating pad_rows?

File: vllm/worker/tt_model_runner.py [hunk 2]
```diff
@@ -370,6 +384,9 @@ def prepare_model_input(
                 # tokens
                 generation_token = seq_data.get_last_token_id()
                 input_tokens_list.append(generation_token)
+                if penalties_requested:
+                    # need prefill tokens to create prompt histogram
+                    decode_prompt_tokens.append(seq_data.get_token_ids())
```
Comment (skhorasganiTT) [inline vllm/worker/tt_model_runner.py]: This variable name (as well as prompt_tokens* vars below) is misleading since it actually stores the full prompt+output tokens. Also currently all penalty types are using prompt+output, but only repetition penalty needs prompt+output while frequency + presence need just output I believe. These should be distinguished - maybe similar to the V1 SamplingMetadata (https://github.com/tenstorrent/vllm/blob/dev/vllm/v1/sample/metadata.py#L13) there should be a separate prompt_token_ids and output_token_ids stored when needed.

Comment (skhorasganiTT) [reply vllm/worker/tt_model_runner.py]: 1 - What if decode is interrupted by a prefill and then resumed? Does the model still retain the histogram? 
2 - Also if we only ever need the prompt tokens then why wouldn't the prefill inputs just be stored in the model?
3 - What about previous comment about only repetition penalty needing prompt+output while frequency + presence need just output - `penalties_requested` currently does not take that into account.

File: vllm/worker/tt_model_runner.py [hunk 3]
```diff
@@ -399,39 +416,49 @@ def prepare_model_input(
...
+                    for key, value in param_bundle.items():
+                        if top_pk_sampling_params[key] != value:
+                            # This should never happen, we always fall back
+                            # to a different sampling implementation if needed
+                            raise ValueError(
+                                "Attempting non-uniform sampling when unsupported"  # noqa: E501
+                            )
+
+        prompt_tokens_tensor: Optional[torch.Tensor] = None
+        if (not is_prompt and penalties_requested and perform_device_sampling):
```
[the remark this answers is not in the document]
Comment (skhorasganiTT) [reply vllm/worker/tt_model_runner.py]: That is a bug on the model side, vLLM should still do the right thing here and pass in the needed inputs for prefill, as it does for the sampling params. Though this is dependent on resolution to other comment above.

File: vllm/worker/tt_model_runner.py [hunk 4]
```diff
@@ -569,6 +625,25 @@ def prepare_model_input(
...
+                                    device="cpu")
+                    ])
+                    req_ids_tensor = torch.cat([
+                        torch.tensor(req_ids),
+                        torch.zeros(batch_pad_len,
+                                    dtype=torch.int32,
+                                    device="cpu")
+                    ])
+                    new_users_tensor = req_ids_tensor != self.prev_req_ids_tensor
+                    self.prev_req_ids_tensor = new_users_tensor
```
[the remark this answers is not in the document]
Comment (skhorasganiTT) [reply vllm/worker/tt_model_runner.py]: order of a user can change if other active users finish but since it's now using get_output_len this is now longer an issue

File: vllm/worker/tt_model_runner.py [hunk 5]
```diff
@@ -370,6 +394,13 @@ def prepare_model_input(
                 # tokens
                 generation_token = seq_data.get_last_token_id()
                 input_tokens_list.append(generation_token)
+                req_ids.append(seq_id)
```
Comment (skhorasganiTT) [inline vllm/worker/tt_model_runner.py]: seq_ids are already stored in seq_groups_list, no need to maintain a separate list

File: vllm/worker/tt_model_runner.py [hunk 6]
```diff
@@ -370,6 +394,13 @@ def prepare_model_input(
...
+                if penalties_requested:
+                    # need prefill tokens to create prompt histogram
+                    decode_prompt_tokens.append(seq_data.prompt_token_ids_array)
```
Comment (skhorasganiTT) [inline vllm/worker/tt_model_runner.py]: Please add a comment here saying that we technically only need prompt tokens for repetition_penalty and the reason it's passed in (automatically masked on metal side)

File: vllm/worker/tt_model_runner.py [hunk 7]
```diff
@@ -399,39 +430,57 @@ def prepare_model_input(
...
+                for key, value in (
+                    ("temperature", sampling_params.temperature),
+                    ("top_k", sampling_params.top_k),
+                    ("top_p", sampling_params.top_p),
+                    ("presence_penalty", sampling_params.presence_penalty),
+                    ("frequency_penalty", sampling_params.frequency_penalty),
+                    ("repetition_penalty", sampling_params.repetition_penalty),
+                    ("seed", sampling_params.seed),
+                ):
+                    top_pk_sampling_params.setdefault(key, []).append(value)
```
Comment (skhorasganiTT) [inline vllm/worker/tt_model_runner.py]: Can simplify by iterating over just keys as done before and appending getattr(sampling_params, key))

File: vllm/worker/tt_model_runner.py [hunk 8]
```diff
@@ -870,6 +968,12 @@ def _execute_model_single_step(self,
             execute_model_kwargs["prompt_lens"] = model_input.prompt_lens
         else:
             execute_model_kwargs["start_pos"] = model_input.input_positions
+            execute_model_kwargs["reset_batch"] = model_input.reset_batch
```
Comment (skhorasganiTT) [inline vllm/worker/tt_model_runner.py]: Can you add a comment here indicating what reset_batch is used for?
